src/System.IO/src/System/IO/BufferedStream.cs

```diff
@@ -0,0 +1,1105 @@
+        private int _writePos;                              // Write pointer within shared buffer.
+        private Task<int> _lastSyncCompletedReadTask;     // The last successful Task returned from ReadAsync
+                                                            // (perf optimization for successive reads of the same size)
+                                                            // Removing a private default constructor is a breaking change for the DataDebugSerializer.
+                                                            // Because this ctor was here previously we need to keep it around.
+        private SemaphoreSlim _asyncActiveSemaphore;
+
+        internal SemaphoreSlim EnsureAsyncActiveSemaphoreInitialized()
+        {
+            // Lazily-initialize _asyncActiveSemaphore.  As we're never accessing the SemaphoreSlim's
+            // WaitHandle, we don't need to worry about Disposing it.
+            return LazyInitializer.EnsureInitialized(ref _asyncActiveSemaphore, () => new SemaphoreSlim(1, 1));
+        }
+
+        private BufferedStream() { }
+
+        public BufferedStream(Stream stream)
+            : this(stream, DefaultBufferSize)
+        {
+        }
+
+        public BufferedStream(Stream stream, int bufferSize)
+        {
+            if (stream == null)
+                throw new ArgumentNullException("stream");
+
+            if (bufferSize <= 0)
+                throw new ArgumentOutOfRangeException("bufferSize", SR.Format(SR.ArgumentOutOfRange_MustBePositive, "bufferSize"));
+
+            _stream = stream;
+            _bufferSize = bufferSize;
+
+            // Allocate _buffer on its first use - it will not be used if all reads
+            // & writes are greater than or equal to buffer size.
+
+            if (!_stream.CanRead && !_stream.CanWrite)
+                throw new ObjectDisposedException(null, SR.ObjectDisposed_StreamClosed);
+        }
+
+        private void EnsureNotClosed()
+        {
+            if (_stream == null)
+                throw new ObjectDisposedException(null, SR.ObjectDisposed_StreamClosed);
+        }
+
+        private void EnsureCanSeek()
+        {
+            Debug.Assert(_stream != null);
+
+            if (!_stream.CanSeek)
+                throw new NotSupportedException(SR.NotSupported_UnseekableStream);
+        }
+
+        private void EnsureCanRead()
+        {
+            Debug.Assert(_stream != null);
+
+            if (!_stream.CanRead)
+                throw new NotSupportedException(SR.NotSupported_UnreadableStream);
+        }
+
+        private void EnsureCanWrite()
+        {
+            Debug.Assert(_stream != null);
+
+            if (!_stream.CanWrite)
+                throw new NotSupportedException(SR.NotSupported_UnwritableStream);
+        }
+
+        private void EnsureShadowBufferAllocated()
+        {
+            Debug.Assert(_buffer != null);
+            Debug.Assert(_bufferSize > 0);
+
+            // Already have shadow buffer?
```

Comment:
Nit: I think my previous confusion may have been that this comment only applies to the first half of the if, not the second half.  I'd suggest augmenting the comment to address the second half as well, e.g. "Already have a shadow buffer? Or is the user-specified buffer size already so large that we don't want to create one?"